### PR TITLE
Add a universal table picker

### PR DESCRIPTION
**Pull request by aschempp**
Replaces https://github.com/contao/core-bundle/pull/1171 and https://github.com/contao/core-bundle/pull/999

It's basically the same implementation as by @qzminski. It also adds a picker for articles and content elements in the respective content elements.

Example configuration:

```php
$GLOBALS['TL_DCA']['tl_module']['fields']['article'] = [
    'label' => &$GLOBALS['TL_LANG']['tl_module']['article'],
    'exclude' => true,
    'foreignKey' => 'tl_article.title',
    'inputType' => 'picker',
    'eval' => [
        //'context' => 'event', // optional, use a custom picker provider
        //'multiple'    => true, // switch to checkbox selection
        //'orderField'     => 'articleOrder',
    ],
    'sql' => ['type' => 'integer', 'unsigned' => true],
];
```

Contrary to https://github.com/contao/core-bundle/pull/1171, the `foreignKey` attribute is used to generate the value from the regular DCA callbacks. If that's not the desired result, implement a custom widget that takes over the rendering.

~FYI this does not mean that any DCA table can be picked, it still always needs a valid picker provider.~ Now added a universal picker provider that can select from any DCA table 🎉

### CODE REVIEW

File: calendar-bundle/src/Resources/config/services.yml [hunk 1]
```diff
@@ -13,3 +13,14 @@ services:
             - '@security.helper'
         tags:
             - { name: contao.picker_provider, priority: 96 }
+
+    contao_calendar.picker.event_content_provider:
```
Comment (leofeyer) [inline calendar-bundle/src/Resources/config/services.yml]: We have agreed on using the class name as service key for all new services, so this needs to become `Contao\CalendarBundle\Picker\EventContentPickerProvider`.

File: calendar-bundle/src/Resources/contao/languages/en/default.xlf [hunk 2]
```diff
@@ -65,6 +65,9 @@
       <trans-unit id="MSC.eventPicker">
         <source>Events</source>
       </trans-unit>
+      <trans-unit id="MSC.eventContentPicker">
+        <source>Events</source>
```
Comment (leofeyer) [inline calendar-bundle/src/Resources/contao/languages/en/default.xlf]: Is there no way to re-use `MSC.eventPicker` here?

File: core-bundle/src/Picker/AbstractContentPickerProvider.php [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+        }
+
+        $qb = $this->connection->createQueryBuilder();
+        $qb
+            ->select('ptable')
+            ->from('tl_content')
+            ->where($qb->expr()->eq('id', $config->getValue()))
+        ;
+
+        if ($qb->execute()->fetchColumn() === $this->getParentTable()) {
```
Comment (leofeyer) [inline core-bundle/src/Picker/AbstractContentPickerProvider.php]: This can be simplified to `return $qb->execute()->fetchColumn() === $this->getParentTable()`, can't it?

File: core-bundle/src/Picker/AbstractContentPickerProvider.php [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+        if ($source = $config->getExtra('source')) {
+            $attributes['preserveRecord'] = $source;
+        }
+
+        if ($value) {
+            $intval = static function ($val) {
+                return (int) $val;
+            };
+
+            $attributes['value'] = array_map($intval, explode(',', $value));
```
Comment (leofeyer) [inline core-bundle/src/Picker/AbstractContentPickerProvider.php]: This should be `array_map('\intval' …`.

File: core-bundle/src/Picker/ArticlePickerProvider.php [hunk 5]
```diff
@@ -70,6 +74,27 @@ public function getDcaAttributes(PickerConfig $config): array
...
+            if ($source = $config->getExtra('source')) {
+                $attributes['preserveRecord'] = $source;
+            }
+
+            if ($value) {
+                $intval = static function ($val) {
+                    return (int) $val;
+                };
+
+                $attributes['value'] = array_map($intval, explode(',', $value));
```
Comment (leofeyer) [inline core-bundle/src/Picker/ArticlePickerProvider.php]: This should be `array_map('\intval' ….`

File: core-bundle/src/Picker/ArticlePickerProvider.php [hunk 6]
```diff
@@ -84,8 +109,12 @@ public function getDcaAttributes(PickerConfig $config): array
     /**
      * {@inheritdoc}
      */
-    public function convertDcaValue(PickerConfig $config, $value): string
+    public function convertDcaValue(PickerConfig $config, $value)
```
Comment (leofeyer) [inline core-bundle/src/Picker/ArticlePickerProvider.php]: Why did you remove the return type?

File: core-bundle/src/Resources/config/services.yml [hunk 7]
```diff
@@ -336,6 +336,17 @@ services:
         tags:
             - { name: contao.picker_provider }
 
+    contao.picker.article_content_provider:
```
Comment (leofeyer) [inline core-bundle/src/Resources/config/services.yml]: We have agreed on using the class name as service key for all new services, so this needs to become `Contao\CoreBundle\Picker\ArticleContentPickerProvider`.

File: core-bundle/src/Resources/contao/dca/tl_content.php [hunk 8]
```diff
@@ -1254,15 +1261,37 @@ public function getArticleAlias(Contao\DataContainer $dc)
 		return $arrAlias;
 	}
 
+	/**
+	 * Saves the article alias
```
Comment (leofeyer) [inline core-bundle/src/Resources/contao/dca/tl_content.php]: This does not seem to match what the method does, does it?

File: core-bundle/src/Resources/contao/dca/tl_content.php [hunk 9]
```diff
@@ -1341,6 +1374,24 @@ public function getAlias()
...
+	 *
+	 * @param mixed         $varValue
+	 * @param DataContainer $dc
+	 *
+	 * @return mixed
+	 */
+	public function saveAlias($varValue, DataContainer $dc)
```
Comment (leofeyer) [inline core-bundle/src/Resources/contao/dca/tl_content.php:1408]: This method seems duplicate?

File: core-bundle/src/Resources/contao/languages/en/default.xlf [hunk 10]
```diff
@@ -1385,6 +1388,9 @@
       <trans-unit id="MSC.articlePicker">
         <source>Articles</source>
       </trans-unit>
+      <trans-unit id="MSC.articleContentPicker">
+        <source>Articles</source>
```
Comment (leofeyer) [inline core-bundle/src/Resources/contao/languages/en/default.xlf]: Can we re-use `MSC.articlePicker` here?

File: news-bundle/src/Resources/config/services.yml [hunk 11]
```diff
@@ -13,3 +13,14 @@ services:
             - '@security.helper'
         tags:
             - { name: contao.picker_provider, priority: 128 }
+
+    contao_news.picker.news_content_provider:
```
Comment (leofeyer) [inline news-bundle/src/Resources/config/services.yml]: We have agreed on using the class name as service key for all new services, so this needs to become `Contao\NewsBundle\Picker\NewsContentPickerProvider`.

File: news-bundle/src/Resources/contao/languages/en/default.xlf [hunk 12]
```diff
@@ -35,6 +35,9 @@
       <trans-unit id="MSC.newsPicker">
         <source>News</source>
       </trans-unit>
+      <trans-unit id="MSC.newsContentPicker">
+        <source>News</source>
```
Comment (leofeyer) [inline news-bundle/src/Resources/contao/languages/en/default.xlf]: Re-using `MSC.newsPicker` anyone?

**Review by leofeyer — CHANGES_REQUESTED** [review]

**Review by Toflar — COMMENTED** [review]
LGTM 👍

File: core-bundle/src/Picker/AbstractContentPickerProvider.php [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+
+            if ($data['ptable'] === $this->getParentTable()) {
+                $params['table'] = 'tl_content';
+                $params['id'] = $data['pid'];
+            }
+        }
+
+        return $params;
+    }
+
```
Comment (qzminski) [inline core-bundle/src/Picker/AbstractContentPickerProvider.php]: A short method description would be nice especially that you use `{@inheritdoc}` in child classes.

File: core-bundle/src/Picker/ArticlePickerProvider.php [hunk 14]
```diff
@@ -84,8 +109,12 @@ public function getDcaAttributes(PickerConfig $config): array
...
     {
+        if ('article' === $config->getContext()) {
+            return (int) $value;
```
Comment (qzminski) [inline core-bundle/src/Picker/ArticlePickerProvider.php]: I think it should return always the same type so convert this one to a string rather than to an integer. Then you could restore the return type above.

File: core-bundle/src/Resources/contao/classes/Ajax.php [hunk 15]
```diff
@@ -328,7 +329,21 @@ public function executePostActions(DataContainer $dc)
...
+				{
+					case 'reloadPicker':
+						$strKey = 'picker';
+						break;
+
+					case 'reloadPagetree':
+						$strKey = 'pageTree';
+						break;
+
+					case 'reloadFiletree':
```
Comment (qzminski) [inline core-bundle/src/Resources/contao/classes/Ajax.php]: No need for it as it fallbacks to default anyway.

File: core-bundle/src/Resources/contao/widgets/Picker.php [hunk 16]
```diff
@@ -0,0 +1,273 @@
...
+
+		return $return;
+	}
+
+	protected function generateValues($blnHasOrder): array
+	{
+		$arrRelations = DcaExtractor::getInstance($this->strTable)->getRelations();
+		$strRelatedTable = $arrRelations[$this->strField]['table'] ?? null;
+
+		if (!$strRelatedTable) {
```
Comment (qzminski) [inline core-bundle/src/Resources/contao/widgets/Picker.php]: CS

**Review by qzminski — DISMISSED** [review]
Looks good to me, just a few minor details that could be improved but without an impact on the picker itself.

File: core-bundle/src/Picker/ArticlePickerProvider.php (hunk 14 above)
Comment (aschempp) [reply core-bundle/src/Picker/ArticlePickerProvider.php]: That's just what all the other pickers to as well 🤷‍♂ 
https://github.com/contao/contao/blob/master/core-bundle/src/Picker/PagePickerProvider.php#L113

File: core-bundle/src/Resources/contao/classes/Ajax.php (hunk 15 above)
Comment (aschempp) [reply core-bundle/src/Resources/contao/classes/Ajax.php]: I like it to be more explicit. One can read in the code what other case there is.

File: core-bundle/src/Picker/AbstractContentPickerProvider.php (hunk 13 above)
Comment (aschempp) [reply core-bundle/src/Picker/AbstractContentPickerProvider.php]: Thanks, fixed in 1ce12911b7031b15d5436aa79e12b8dc87caedbe

File: core-bundle/src/Resources/contao/widgets/Picker.php (hunk 16 above)
Comment (aschempp) [reply core-bundle/src/Resources/contao/widgets/Picker.php]: Thanks, fixed in 1ce12911b7031b15d5436aa79e12b8dc87caedbe

**Comment by leofeyer** [comment]
As briefly discussed in Slack with @aschempp, we could add an universal picker provider for all pickers that return a database ID.

File: core-bundle/src/Picker/AbstractContentPickerProvider.php (hunk 4 above)
Comment (aschempp) [reply core-bundle/src/Picker/AbstractContentPickerProvider.php]: jup, no idea why thats present in the other pickers. Fine with me to change 😂

File: calendar-bundle/src/Resources/contao/languages/en/default.xlf (hunk 2 above)
Comment (aschempp) [reply calendar-bundle/src/Resources/contao/languages/en/default.xlf]: Not sure 🤷‍♂ it's a different picker, so it should have a different label. Then again, the tab/label actually represents the back end module, not the picker implementation…

File: core-bundle/src/Picker/ArticlePickerProvider.php (hunk 6 above)
Comment (aschempp) [reply core-bundle/src/Picker/ArticlePickerProvider.php]: because it can be an integer as well now

File: core-bundle/src/Resources/contao/dca/tl_content.php (hunk 9 above)
Comment (aschempp) [reply core-bundle/src/Resources/contao/dca/tl_content.php:1408]: It's not, it's `$dc->activeRecord->pid` vs. `$dc->activeRecord->id`

File: core-bundle/src/Resources/contao/languages/en/default.xlf (hunk 10 above)
Comment (aschempp) [reply core-bundle/src/Resources/contao/languages/en/default.xlf]: 🤷‍♂

**Comment by aschempp** [comment]
@leofeyer I think we can merge this if there's no more feedback. I'll create a separate PR for the universal picker provider.

File: core-bundle/src/Picker/AbstractContentPickerProvider.php (hunk 4 above)
Comment (aschempp) [reply core-bundle/src/Picker/AbstractContentPickerProvider.php]: Changed in b1eeebb30a3fa26f5bc1c042e91ed6889585acb1

File: core-bundle/src/Picker/ArticlePickerProvider.php (hunk 5 above)
Comment (aschempp) [reply core-bundle/src/Picker/ArticlePickerProvider.php]: Changed in b1eeebb30a3fa26f5bc1c042e91ed6889585acb1

File: core-bundle/src/Resources/contao/dca/tl_content.php (hunk 8 above)
Comment (aschempp) [reply core-bundle/src/Resources/contao/dca/tl_content.php]: Changed in b1eeebb30a3fa26f5bc1c042e91ed6889585acb1

File: core-bundle/src/Picker/AbstractContentPickerProvider.php (hunk 3 above)
Comment (aschempp) [reply core-bundle/src/Picker/AbstractContentPickerProvider.php]: Changed in b1eeebb30a3fa26f5bc1c042e91ed6889585acb1

**Comment by aschempp** [comment]
PR updated with a universal picker provider to pick any DC record 🎉

File: core-bundle/src/Picker/UniversalPickerProvider.php [hunk 17]
```diff
@@ -0,0 +1,267 @@
...
+     * {@inheritdoc}
+     */
+    public function getUrl(PickerConfig $config): string
+    {
+        $table = $this->getTable($config->getContext());
+        $modules = $this->getModulesForTable($config->getContext());
+        $value = $config->getValue();
+
+        if (!$value) {
+            return $this->getUrlForValue($config, array_keys($modules)[0]);
```
Comment (Toflar) [inline core-bundle/src/Picker/UniversalPickerProvider.php]: The case where `$modules` is empty should be handled even though that is very unlikely to happen :)

File: core-bundle/src/Picker/UniversalPickerProvider.php [hunk 18]
```diff
@@ -0,0 +1,267 @@
...
+        }
+
+        // Use the first value if array to find a database record
+        $value = array_map('\intval', explode(',', $value))[0];
+
+        Controller::loadDataContainer($table);
```
Comment (Toflar) [inline core-bundle/src/Picker/UniversalPickerProvider.php]: Should use the `Adapter` for tests.

File: core-bundle/src/Picker/UniversalPickerProvider.php [hunk 19]
```diff
@@ -0,0 +1,267 @@
...
+            $qb->select('pid')->from($table)->where($qb->expr()->eq('id', $value));
+
+            if ($dynamicPtable) {
+                $qb->addSelect('ptable');
+            }
+
+            $data = $qb->execute()->fetch();
+
+            if (false === $data) {
+                return $this->getUrlForValue($config, array_keys($modules)[0]);
```
Comment (Toflar) [inline core-bundle/src/Picker/UniversalPickerProvider.php]: Same here, what if `$modules` is empty?

File: core-bundle/src/Picker/UniversalPickerProvider.php [hunk 20]
```diff
@@ -0,0 +1,267 @@
...
+            }
+
+            $pid = $data['pid'];
+
+            if ($dynamicPtable) {
+                $ptable = $data['ptable'] ?: $ptable;
+
+                // Backwards compatibility, assuming old data in tl_content
+                if (!$ptable) {
```
Comment (Toflar) [inline core-bundle/src/Picker/UniversalPickerProvider.php]: This is the third nested if statement. Let's refactor this a bit to have more but simpler methods.

File: core-bundle/src/Picker/UniversalPickerProvider.php [hunk 21]
```diff
@@ -0,0 +1,267 @@
...
+
+        return $menu->getFirstChild();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function supportsContext($context): bool
+    {
+        return 0 === strpos($context, 'universal.') && 0 !== \count($this->getModulesForTable($context));
```
Comment (Toflar) [inline core-bundle/src/Picker/UniversalPickerProvider.php]: `universal.` should probably be a private class constant :)

File: core-bundle/src/Resources/contao/widgets/Picker.php [hunk 22]
```diff
@@ -0,0 +1,281 @@
...
+	protected $strTemplate = 'be_widget';
+
+	/**
+	 * Load the database object
+	 *
+	 * @param array $arrAttributes
+	 */
+	public function __construct($arrAttributes=null)
+	{
+		$this->import(Database::class, 'Database');
```
Comment (Toflar) [inline core-bundle/src/Resources/contao/widgets/Picker.php:37]: Really? 🙄

**Review by Toflar — CHANGES_REQUESTED** [review]

File: core-bundle/src/Resources/contao/widgets/Picker.php (hunk 22 above)
Comment (aschempp) [reply core-bundle/src/Resources/contao/widgets/Picker.php:37]: It's legacy code 🤷‍♂

File: core-bundle/src/Picker/UniversalPickerProvider.php (hunk 19 above)
Comment (aschempp) [reply core-bundle/src/Picker/UniversalPickerProvider.php]: Added exception in 319ef87f32e7fc220b67778d566e6a6e8fd91b35

File: core-bundle/src/Picker/UniversalPickerProvider.php (hunk 17 above)
Comment (aschempp) [reply core-bundle/src/Picker/UniversalPickerProvider.php]: Added exception in 319ef87f32e7fc220b67778d566e6a6e8fd91b35

File: core-bundle/src/Picker/UniversalPickerProvider.php (hunk 20 above)
Comment (aschempp) [reply core-bundle/src/Picker/UniversalPickerProvider.php]: Refactored in 319ef87f32e7fc220b67778d566e6a6e8fd91b35

File: core-bundle/src/Picker/UniversalPickerProvider.php (hunk 21 above)
Comment (aschempp) [reply core-bundle/src/Picker/UniversalPickerProvider.php]: Refactored in 319ef87f32e7fc220b67778d566e6a6e8fd91b35

File: core-bundle/src/Picker/UniversalPickerProvider.php (hunk 18 above)
Comment (aschempp) [reply core-bundle/src/Picker/UniversalPickerProvider.php]: Refactored in 319ef87f32e7fc220b67778d566e6a6e8fd91b35

File: core-bundle/src/Picker/UniversalPickerProvider.php [hunk 23]
```diff
@@ -0,0 +1,294 @@
...
+     */
+    public function isCurrent(PickerConfig $config): bool
+    {
+        return 0 === strpos($config->getCurrent(), self::PREFIX);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getDcaTable(/* PickerConfig $config */): string
```
Comment (leofeyer) [inline core-bundle/src/Picker/UniversalPickerProvider.php]: If you made the argument optional, you could add it as `getDcaTable(PickerConfig $config = null)`. See https://3v4l.org/Rr2UF.

File: core-bundle/src/Picker/PickerMenuInterface.php [hunk 24]
```diff
@@ -0,0 +1,20 @@
...
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Picker;
+
+use Knp\Menu\ItemInterface;
+
+interface PickerMenuInterface
+{
+    public function addMenuItems(ItemInterface $menu, PickerConfig $config): void;
```
Comment (leofeyer) [inline core-bundle/src/Picker/PickerMenuInterface.php:22]: We should add a comment here describing what the method does.

File: core-bundle/src/Picker/UniversalPickerProvider.php [hunk 25]
```diff
@@ -0,0 +1,294 @@
...
+        if ($ptable) {
+            foreach ($modules as $k => $tables) {
+                if (\in_array($ptable, $tables, true)) {
+                    $module = $k;
+                    break;
+                }
+            }
+        }
+
+        if (0 === array_search($table, $modules[$module], true)) {
```
Comment (leofeyer) [inline core-bundle/src/Picker/UniversalPickerProvider.php]: What exactly does this line do?

File: core-bundle/src/Picker/UniversalPickerProvider.php [hunk 26]
```diff
@@ -0,0 +1,294 @@
...
+
+        return $attributes;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function convertDcaValue(PickerConfig $config, $value)
+    {
+        return (int) $value;
```
Comment (leofeyer) [inline core-bundle/src/Picker/UniversalPickerProvider.php]: This implies that we can only pick IDs. What about UUIDs?

File: core-bundle/src/Resources/contao/dca/tl_content.php [hunk 27]
```diff
@@ -730,38 +730,36 @@
...
+			'inputType'               => 'picker',
+			'eval'                    => array('mandatory'=>true, 'tl_class'=>'clr'),
+			'save_callback' => array
 			(
-				array('tl_content', 'editAlias')
+				array('tl_content', 'saveAlias'),
 			),
-			'sql'                     => "int(10) unsigned NOT NULL default 0"
+			'sql'                     => "int(10) unsigned NOT NULL default 0",
+			'relation'                => array('type'=>'hasOne', 'load'=>'lazy', 'table'=>'tl_content')
```
Comment (leofeyer) [inline core-bundle/src/Resources/contao/dca/tl_content.php:740]: How does this work if there is no foreign key?

Comment (aschempp) [reply core-bundle/src/Resources/contao/dca/tl_content.php:740]: You either need `foreignKey` or a table relation. If you have neither, the button in the picker will be disabled so you can't pick anything (simply because there is no picker provider for your context).

File: core-bundle/src/Picker/UniversalPickerProvider.php (hunk 26 above)
Comment (aschempp) [reply core-bundle/src/Picker/UniversalPickerProvider.php]: There's a separate file picker, this should only be for DC_Table imho.

File: core-bundle/src/Picker/UniversalPickerProvider.php (hunk 25 above)
Comment (aschempp) [reply core-bundle/src/Picker/UniversalPickerProvider.php]: If the table is the first in the module, we don't need to add `table=xy` to the URL.

File: core-bundle/src/Picker/UniversalPickerProvider.php (hunk 23 above)
Comment (aschempp) [reply core-bundle/src/Picker/UniversalPickerProvider.php]: No, the class that implements the interface must have the exact same arguments. This would break existing picker providers.

Comment (aschempp) [reply core-bundle/src/Picker/UniversalPickerProvider.php]: Or you mean we can add it to the picker, but not to the interface?

File: core-bundle/src/Picker/PickerMenuInterface.php (hunk 24 above)
Comment (aschempp) [reply core-bundle/src/Picker/PickerMenuInterface.php:22]: Added in 03ca7e3a958926344dec1d8c7277cbf4ad10ef80

File: core-bundle/src/Picker/UniversalPickerProvider.php (hunk 23 above)
Comment (aschempp) [reply core-bundle/src/Picker/UniversalPickerProvider.php]: Changed in 03ca7e3a958926344dec1d8c7277cbf4ad10ef80

**Review by Toflar — COMMENTED** [review]
LGTM except for missing tests. I think they might influence the outcome again so I'll approve once this PR is completely ready.

File: core-bundle/src/Picker/UniversalPickerProvider.php (hunk 26 above)
Comment (leofeyer) [reply core-bundle/src/Picker/UniversalPickerProvider.php]: It is called `UniversalPickerProvider` though, which implies that it is universal, doesn't it? 😄 Maybe we should rename it to `DatabasePickerProvider` then?

File: core-bundle/src/Picker/UniversalPickerProvider.php (hunk 25 above)
Comment (leofeyer) [reply core-bundle/src/Picker/UniversalPickerProvider.php]: I have added your explanation as comment in 8214ec3d347ceaca42ab2c04fa4df928e0c997fb.

File: core-bundle/src/Picker/UniversalPickerProvider.php (hunk 26 above)
Comment (aschempp) [reply core-bundle/src/Picker/UniversalPickerProvider.php]: It theoretically picks anything from a DataContainer, but it actually only supports DC_Table. Not sure how to name that, not sure if it's really relevant either 😂

Comment (aschempp) [reply core-bundle/src/Picker/UniversalPickerProvider.php]: Maybe it should be a `TablePickerProvider`?

Comment (leofeyer) [reply core-bundle/src/Picker/UniversalPickerProvider.php]: I like `TablePickerProvider`. However, I would like it even more, if it remained a `UniversalPickerProvider`, which is really universal. 😉

Comment (aschempp) [reply core-bundle/src/Picker/UniversalPickerProvider.php]: That's never gonna work imho. It _might_ pick multiple DCs, but only the ones we know about. I think we're better of with a `DC_Table` picker, I'll adjust accordingly 😊

Comment (aschempp) [reply core-bundle/src/Picker/UniversalPickerProvider.php]: Changed in c759e3c8d45537527d783604e83d1d86d76e0090
I tried implementing a picker for products in Isotope, but there's a lot of other things failing on that (e.g. the DC_ProductData is still based on DC_Table from Contao 3, so there's no picker support). But I found and fixed some other issues in the process.

**Review by leofeyer — CHANGES_REQUESTED** [review]
I like the latest changes, however, the `AbstractDataContainerPickerProvider` class does not really make sense to me. It contains a lot of code that is only required for `DC_Table` and that would have to be overwritten to make it work with other DCAs.

I would prefer not to have an abstract class. Just rename it to `TablePickerProvider` and we are all set.

**Comment by aschempp** [comment]
> I like the latest changes, however, the `AbstractDataContainerPickerProvider` class does not really make sense to me. It contains a lot of code that is only required for `DC_Table` and that would have to be overwritten to make it work with other DCAs.

How come you think that? Every custom DC I ever implemented did extend DC_Table and adjusted some fields. I don't think it's better to extend the `TablePickerProvider`, which would be the case if there's no abstract.

**Comment by leofeyer** [comment]
You should not extend the `TablePickerProvider` for a folder picker provider obviously.

An abstract class should contain code that is used by 90% of their child classes. In this case, however, only about 20% of the code is relevant for a folder picker, which indicates that the abstract class is too concrete. 🤷‍♂

**Comment by aschempp** [comment]
> An abstract class should contain code that is used by 90% of their child classes. In this case, however, only about 20% of the code is relevant for a folder picker, which indicates that the abstract class is too concrete. 🤷‍♂

The only method that's possibly not related to DC_Folder is the database query, but any other table will most certainly use the database, otherwise it would not make sense to use be a DataContainer picker. Also, there's already a picker for files, so there will never be a picker for DC_Folder 😉

**Comment by leofeyer** [comment]
A DataContainer can be anything and does not necessarily have to use the database (e.g. `DC_Files`).

**Comment by aschempp** [comment]
Here's another attempt at class naming, @leofeyer is that better? Also added full code coverage 😎

**Review by Toflar — DISMISSED** [review]
Cannot say if it works as supposed to but I approve the implementation :) 
It will maybe need some tweaks in terms of how the picker displays the entries but that's something we can still polish in the RC phase with some feedback of the users etc.

**Review by leofeyer — APPROVED** [review]

**Review by Toflar — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @aschempp.

**Comment by fritzmg** [comment]
@aschempp I am trying to gather some documentation for this. However I noticed some discrepancies with the original description in this issue. If I try to implement a news picker like this:

```php
// contao/dca/tl_content.php
use Contao\CoreBundle\DataContainer\PaletteManipulator;

$GLOBALS['TL_DCA']['tl_content']['fields']['news'] = [
    'label' => ['News', 'Choose a news article.'],
    'foreignKey' => 'tl_news.headline',
    'inputType' => 'picker',
    'eval' => ['tl_class' => 'clr'],
    'sql' => ['type' => 'integer', 'unsigned' => true, 'default' => 0],
];

PaletteManipulator::create()
    ->addField('news', 'text', PaletteManipulator::POSITION_AFTER)
    ->applyToPalette('text', 'tl_content')
;
```

Then the following error will occur:

```
Exception:
The table name must not be empty

  at vendor\contao\contao\core-bundle\src\Resources\contao\library\Contao\DcaLoader.php:50
  at Contao\DcaLoader->__construct()
  at ReflectionClass->newInstanceArgs()
     (vendor\contao\contao\core-bundle\src\Framework\ContaoFramework.php:177)
  at Contao\CoreBundle\Framework\ContaoFramework->createInstance()
     (vendor\contao\contao\core-bundle\src\Picker\AbstractTablePickerProvider.php:133)
  at Contao\CoreBundle\Picker\AbstractTablePickerProvider->supportsContext()
     (vendor\contao\contao\core-bundle\src\Picker\PickerBuilder.php:94)
  at Contao\CoreBundle\Picker\PickerBuilder->supportsContext()
     (vendor\contao\contao\core-bundle\src\Resources\contao\widgets\Picker.php:138)
  at Contao\Picker->generate()
     (vendor\contao\contao\core-bundle\src\Resources\contao\library\Contao\Widget.php:653)
  at Contao\Widget->generateWithError()
     (vendor\contao\contao\core-bundle\src\Resources\contao\templates\backend\be_widget.html5:3)
  at include('vendor/contao/contao/core-bundle/src/Resources/contao/templates/backend/be_widget.html5')
     (vendor\contao\contao\core-bundle\src\Resources\contao\library\Contao\TemplateInheritance.php:100)
  at Contao\Widget->inherit()
     (vendor\contao\contao\core-bundle\src\Resources\contao\library\Contao\Widget.php:600)
  at Contao\Widget->parse()
     (vendor\contao\contao\core-bundle\src\Resources\contao\classes\DataContainer.php:670)
  at Contao\DataContainer->row()
     (vendor\contao\contao\core-bundle\src\Resources\contao\drivers\DC_Table.php:2003)
  at Contao\DC_Table->edit()
     (vendor\contao\contao\core-bundle\src\Resources\contao\classes\Backend.php:648)
  at Contao\Backend->getBackendModule()
     (vendor\contao\contao\core-bundle\src\Resources\contao\controllers\BackendMain.php:167)
  at Contao\BackendMain->run()
     (vendor\contao\contao\core-bundle\src\Controller\BackendController.php:49)
  at Contao\CoreBundle\Controller\BackendController->mainAction()
     (vendor\symfony\http-kernel\HttpKernel.php:157)
  at Symfony\Component\HttpKernel\HttpKernel->handleRaw()
     (vendor\symfony\http-kernel\HttpKernel.php:79)
  at Symfony\Component\HttpKernel\HttpKernel->handle()
     (vendor\symfony\http-kernel\Kernel.php:195)
  at Symfony\Component\HttpKernel\Kernel->handle()
     (web\index.php:31)
  at require('web/index.php')
     (web\app.php:4)
```

However if I then add a relation with `table` then everything is fine. I can even remove `foreignKey` - in fact the `foreignKey` declaration does not seem to have an effect whatsoever.

```php
// contao/dca/tl_content.php
use Contao\CoreBundle\DataContainer\PaletteManipulator;

$GLOBALS['TL_DCA']['tl_content']['fields']['news'] = [
    'label' => ['News', 'Choose a news article.'],
    'inputType' => 'picker',
    'eval' => ['tl_class' => 'clr'],
    'sql' => ['type' => 'integer', 'unsigned' => true, 'default' => 0],
    'relation' => ['type' => 'hasOne', 'load' => 'lazy', 'table' => 'tl_news'],
];

PaletteManipulator::create()
    ->addField('news', 'text', PaletteManipulator::POSITION_AFTER)
    ->applyToPalette('text', 'tl_content')
;
```

Is this correct?

**Comment by aschempp** [comment]
hmm. Did you build the internal cache? The relation table should be taken from the `DcaExtractor` data not the DCA, but that might be a bug …

**Comment by fritzmg** [comment]
No, the error was taken from the `dev` environment, without an internal cache being built.

**Comment by aschempp** [comment]
Can you try if it works with the cache? Then maybe that's the problem…

Anyway, looks like it deserves a bug report to me 😉

**Comment by fritzmg** [comment]
Right, but seeing as it works just with the relation... what advantages would the `foreignKey` definiton bring? Is it supposed to work with either or?

**Comment by aschempp** [comment]
The table relations fall back to foreignKey, so our picker should do that as well.
